src/coprocessor/cache.rs

```diff
@@ -0,0 +1,491 @@
+        DistSQLCache {
+            max_size: capacity,
+            ..Default::default()
+        }
+    }
+
+    pub fn put(&mut self, region_id: u64, k: DistSQLCacheKey, version: u64, res: Vec<u8>) {
+        if !self.is_region_cache_enabled(region_id) {
+            return;
+        }
+
+        let key_size = k.len();
+        if self.map.contains_key(&k) {
+            let mut entry = self.map.get_mut(&k).unwrap();
+            let size_diff = entry.update(region_id, version, key_size, res);
+            self.size += size_diff;
+        } else {
+            let entry = DistSQLCacheEntry::new(region_id, version, key_size, res);
+            self.size += entry.size;
+            self.map.insert(k.clone(), entry);
+            self.update_regions(region_id, k);
+        }
+
+        // Remove entry untile cache size is less or equals than capacity
+        while self.size() > self.capacity() {
+            self.remove_lru();
+        }
+
+        CORP_DISTSQL_CACHE_SIZE_GAUGE_VEC
+            .with_label_values(&["size"])
+            .set(self.size() as f64);
+
+        CORP_DISTSQL_CACHE_SIZE_GAUGE_VEC
+            .with_label_values(&["count"])
+            .set(self.len() as f64);
+    }
+
+    fn check_evict_key(&mut self, region_id: u64, k: &str) {
+        let mut need_remove = false;
+        if let Some(entry) = self.map.get(k) {
+            if let Some(rentry) = self.regions.get(&region_id) {
+                // Region's version is not same as cached evict it
+                if entry.version != rentry.version {
+                    need_remove = true;
+                }
+            }
+        }
+        if need_remove {
+            self.remove(k);
+        }
+    }
+
+    pub fn get_region_version_and_cache_entry(
+        &mut self,
+        region_id: u64,
+        k: &str,
+    ) -> (u64, Option<&Vec<u8>>) {
+        (self.get_region_version(region_id), self.get(region_id, k))
+    }
+
+    pub fn get_region_version(&self, region_id: u64) -> u64 {
+        match self.regions.get(&region_id) {
+            None => 0,
+            Some(item) => item.version,
+        }
+    }
+
+    pub fn get(&mut self, region_id: u64, k: &str) -> Option<&Vec<u8>> {
+        if !self.is_region_cache_enabled(region_id) {
+            return None;
+        }
+
+        self.check_evict_key(region_id, k);
+        if let Some(entry) = self.map.get_refresh(k) {
+            Some(&entry.result)
+        } else {
+            None
+        }
+    }
+
+    pub fn remove(&mut self, k: &str) {
+        let regions = &mut self.regions;
+        if let Some(entry) = self.map.remove(k) {
+            let region_id: u64 = entry.region_id;
+            let mut need_remove_region = false;
+            if let Some(node) = regions.get_mut(&region_id) {
+                // Delete from region cache entry list
+                node.cached_items.remove(k);
+                if !node.cached_items.is_empty() && node.version != 1 {
```

Comment:
what does `node.version == 0` and `node.version == 1` means?

Reply:
I think we should not delete any items in self.regions.